src/main/java/uk/gov/pay/adminusers/service/InviteService.java

```diff
@@ -141,16 +151,32 @@ public void generateOtp(InviteOtpRequest inviteOtpRequest) {
     }
 
     @Transactional
-    public User createInvitedUser(InviteValidateOtpRequest inviteValidateOtpRequest) {
+    public User createInvitedUserAndInvalidateInvite(InviteValidateOtpRequest inviteValidateOtpRequest) {
         Optional<InviteEntity> inviteOptional = inviteDao.findByCode(inviteValidateOtpRequest.getCode());
         if (inviteOptional.isPresent()) {
-            if (!secondFactorAuthenticator.authorize(inviteOptional.get().getOtpKey(), inviteValidateOtpRequest.getOtpCode())) {
-                throw invalidOtpAuthCodeInviteException(inviteValidateOtpRequest.getCode());
-            }
             InviteEntity inviteEntity = inviteOptional.get();
+            if (!secondFactorAuthenticator.authorize(inviteEntity.getOtpKey(), inviteValidateOtpRequest.getOtpCode())) {
+                // "failed login attempt" logic
+                inviteEntity.setLoginCounter(inviteEntity.getLoginCounter() + 1);
+                inviteEntity.setDisabled(inviteEntity.getLoginCounter() >= loginAttemptCap);
+                inviteDao.merge(inviteEntity);
+                // throw UNAUTHORIZED / 401 exception if the otp validation failed
+                throw invalidOtpAuthCodeInviteException(inviteEntity.getCode());
```

Comment:
Are you sure this is working? You have to be very careful when throwing exceptions from a `@Transactional` method. In this case I can almost guarantee the counter logic will not work even if you call `merge` method because you are throwing an exception. I expect the transaction to rollback here and not update the counter. Have a look